paasta_tools/deployd/queue.py

```diff
@@ -146,47 +173,50 @@ def _lock_and_get_entry(self, entry_node: str):
             self.client.delete(lock_path)
             return None
 
-    def _consume(self, entry_node: str):
+    def _consume(self, entry_node: str) -> None:
         # necessary in case we lose connection at some point
         if not self._holds_lock(entry_node):
             return  # TODO: log?
         with self.client.transaction() as transaction:
             transaction.delete(f"{self.locks_path}/{entry_node}")
             transaction.delete(f"{self.entries_path}/{entry_node}")
 
-    def _holds_lock(self, entry_node: str):
+    def _holds_lock(self, entry_node: str) -> bool:
         lock_path = f"{self.locks_path}/{entry_node}"
         self.client.sync(lock_path)
         value, stat = self.client.retry(self.client.get, lock_path)
         return value == self.id
 
-    def _release(self, entry_node: str):
+    def _release(self, entry_node: str) -> None:
         if not self._holds_lock(entry_node):
             return
         self.client.delete(f"{self.locks_path}/{entry_node}")
 
-    def get_available_service_instances(
+    def _get_all_unlocked_service_instances(
         self
-    ) -> Iterable[Tuple[float, ServiceInstance]]:
+    ) -> Iterable[Tuple[float, float, ServiceInstance]]:
         results = []
         with self.local_state_condition:
             for entry_node in self.entry_nodes:
                 if entry_node not in self.locked_entry_nodes:
                     deadline, wait_until = self._parse_entry_node(entry_node)
-                    now = time.time()
-                    if wait_until <= now:
-                        results.append((deadline, entry_node))
+                    data, _ = self.client.get(f"{self.entries_path}/{entry_node}")
```

Comment:
Hm, you're right, we don't actually use the ServiceInstance right now in the metrics; we're just counting numbers.

Honestly I just did it to match what the in-memory queue was providing, and didn't think about what metrics.py was actually using. Maybe I can make a parameter for whether to actually include the ServiceInstance?

Reply:
Yeah I am not sure how much it matter, but I guess I've heard a couple things recently about issues with load on ZK so I figured I'd point it out.  Seems like it's possible that this would slow things down.

I can imagine that in the future we could build some nice monitoring tools that actually fetch the ServiceInstances, but maybe those things could take the list of queue nodes and do their own fetching?